opencensus/exporters/stats/prometheus/internal/prometheus_utils.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "opencensus/exporters/stats/prometheus/internal/prometheus_utils.h"
+
+#include <cctype>
+#include <cstdint>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "absl/time/time.h"
+#include "metrics.pb.h"
+#include "opencensus/stats/stats.h"
+
+namespace opencensus {
+namespace exporters {
+namespace stats {
+
+namespace {
+
+// Replaces non-alphanumeric characters with underscores to satisfy
+// Stackdriver's name requirements.
+std::string SanitizeName(absl::string_view name) {
+  std::string sanitized(name);
+  std::replace_if(sanitized.begin(), sanitized.end(),
+                  [](char c) { return !::isalnum(c); }, '_');
+  return sanitized;
+}
+
+io::prometheus::client::MetricType MetricType(
+    opencensus::stats::Aggregation::Type type) {
+  switch (type) {
+    case opencensus::stats::Aggregation::Type::kCount:
+      return io::prometheus::client::MetricType::COUNTER;
+    case opencensus::stats::Aggregation::Type::kSum:
+      return io::prometheus::client::MetricType::UNTYPED;
+    case opencensus::stats::Aggregation::Type::kDistribution:
+      return io::prometheus::client::MetricType::HISTOGRAM;
+  }
+}
+
+void SetValue(double value, io::prometheus::client::Metric* metric) {
+  metric->mutable_untyped()->set_value(value);
+}
+void SetValue(int64_t value, io::prometheus::client::Metric* metric) {
+  metric->mutable_counter()->set_value(value);
+}
+void SetValue(const opencensus::stats::Distribution& value,
+              io::prometheus::client::Metric* metric) {
+  auto* histogram = metric->mutable_histogram();
+  histogram->set_sample_count(value.count());
+  histogram->set_sample_sum(value.count() * value.mean());
+
+  int64_t cumulative_count = 0;
+  for (int i = 0; i < value.bucket_boundaries().num_buckets(); ++i) {
+    cumulative_count += value.bucket_counts()[i];
+    auto* bucket = histogram->add_bucket();
+    bucket->set_cumulative_count(cumulative_count);
+    // We use lower boundaries plus an underflow bucket; Prometheus uses upper
+    // boundaries, including a +Inf boundary.
```

Comment:
`internal/prometheus_test_server.cc` does `BucketBoundaries::Explicit({0, 10})`

This code produces:
```
bucket{key1="v1",key2="",le="0.000000"} 0 1519637070515
bucket{key1="v1",key2="",le="10.000000"} 12 1519637070515
bucket{key1="v1",key2="",le="+Inf"} 18 1519637070515
```

Shouldn't it be just two buckets? Upper inclusive bounds 10 and +Inf?

Reply:
Internally, we make three buckets out of `BucketBoundaries::Explicit({0, 10})`--an underflow bucket for `[-Inf, 0)`, and buckets `[0, 10)` and `[10, +Inf]`--for Prometheus we report the upper boundaries of those buckets. (Prometheus's rules prevent dropping the underflow bucket--we cannot drop its samples from the sum/mean, and since the bucket counts must equal the overall count we cannot omit any of the buckets without attributing its contents to a different bucket.)

Reply:
That makes sense but wasn't obvious to me.

Is this documented somewhere and I missed it? If not, please document it in a later PR. :)

Reply:
There is a brief comment in prometheus_utils.cc. I should write an md for each of the exporters documenting data model translations in greater depth.